### PR TITLE
[tests-only][full-ci]Added test for getting user information by admin or normal user using Graph API

### PR DESCRIPTION
**Pull request by SagarGi**
### Description

Added Scenarios:

1. `admin user gets the information of a user`
2. `non-admin user tries to get the information of a user`
3. `admin user gets all users`
4. `non-admin user tries get all user`
5. `admin user gets the drive information of a user`
6. `normal user gets his/her own drive information`

### Related Issue

https://github.com/owncloud/ocis/issues/4937

### CODE REVIEW

File: tests/TestHelpers/GraphHelper.php [hunk 1]
```diff
@@ -26,16 +26,43 @@ private static function getRequestHeaders(): array {
...
+	}
+
 	/**
 	 * @param string $id
 	 *
 	 * @return int (1 = true | 0 = false)
 	 */
 	public static function isUUIDv4(string $id): int {
-		$regex = '/^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i';
+		$regex = "/^" . self::getUUIDv4Regex() . "/i";
```
Comment (saw-jan) [inline tests/TestHelpers/GraphHelper.php:43]:
```suggestion
		$regex = "/^" . self::getUUIDv4Regex() . "$/i";
```

File: tests/TestHelpers/GraphHelper.php [hunk 2]
```diff
@@ -26,16 +26,43 @@ private static function getRequestHeaders(): array {
...
 		return preg_match($regex, $id);
 	}
 
+	/**
+	 * @param string $spaceId
+	 *
+	 * @return int (1 = true | 0 = false)
+	 */
+	public static function isSpaceId(string $spaceId): int {
+		$regex = "/^" . self::getUUIDv4Regex() . '\\$' . self::getUUIDv4Regex() . "/i";
```
Comment (saw-jan) [inline tests/TestHelpers/GraphHelper.php:53]:
```suggestion
		$regex = "/^" . self::getUUIDv4Regex() . '\\$' . self::getUUIDv4Regex() . "$/i";
```

File: tests/TestHelpers/GraphHelper.php [hunk 3]
```diff
@@ -26,16 +26,43 @@ private static function getRequestHeaders(): array {
...
+	}
+
+	/**
+	 * @param string $id
+	 *
+	 * @return string
+	 */
+	public static function isUUIDv44(string $id): string {
+		return "hello";
+	}
```
Comment (saw-jan) [inline tests/TestHelpers/GraphHelper.php:64]: hi :smile: 
```suggestion
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  As an admin
+  I want to be able to retrieve user information
+  So that I can see the information
+
+  Background:
+    Given user "Alice" has been created with default attributes and without skeleton files
+    And the administrator has given "Alice" the role "Admin" using the settings api
+
+
+  Scenario: admin user tries get information of a user
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:12]:
```suggestion
  Scenario: admin user gets the information of a user
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    Given user "Brian" has been created with default attributes and without skeleton files
+    When user "Alice" tries to get information of user "Brian" using Graph API
+    Then the HTTP status code should be "200"
+    And the user retrieve API response should contain the following information:
+      | displayName  | id        | mail              | onPremisesSamAccountName |
+      | Brian Murphy | %uuid_v4% | brian@example.org | Brian                    |
+
+
+  Scenario: non-admin user tries get information of a user
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:21]:
```suggestion
  Scenario: non-admin user tries to get the information of a user
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    Given user "Brian" has been created with default attributes and without skeleton files
+    When user "Brian" tries to get information of user "Alice" using Graph API
+    Then the HTTP status code should be "200"
+    And the last response should be an unauthorized response
+
+
+  Scenario: admin user tries get all user
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:28]:
```suggestion
  Scenario: admin user gets all users
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    Then the HTTP status code should be "200"
+    And the user retrieve API response should contain the following information:
+      | displayName  | id        | mail              | onPremisesSamAccountName |
+      | Brian Murphy | %uuid_v4% | brian@example.org | Brian                    |
+
+
+  Scenario: non-admin user tries get information of a user
+    Given user "Brian" has been created with default attributes and without skeleton files
+    When user "Brian" tries to get information of user "Alice" using Graph API
+    Then the HTTP status code should be "200"
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:24]: Won't it be `401`?
```suggestion
    Then the HTTP status code should be "401"
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    When user "Alice" tries to get all user using the Graph API
+    Then the HTTP status code should be "200"
+    And the API response should contain all user with following information:
+      | displayName  | id        | mail              | onPremisesSamAccountName |
+      | Brian Murphy | %uuid_v4% | brian@example.org | Brian                    |
+      | David Lopez  | %uuid_v4% | david@example.org | David                    |
+      | Carol King   | %uuid_v4% | carol@example.org | Carol                    |
+
+
+  Scenario: non-admin user tries get all user
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:43]:
```suggestion
  Scenario: non-admin user tries to get all users
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      | username |
+      | Brian    |
+      | Carol    |
+      | David    |
+    When user "Brian" tries to get all user using the Graph API
+    Then the HTTP status code should be "401"
+    And the last response should be an unauthorized response
+
+
+  Scenario: admin user tries to get drive information of a user
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:54]:
```suggestion
  Scenario: admin user gets the drive information of a user
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+      | id                | %space_id%                       |
+      | name              | Brian Murphy                     |
+      | owner@@@user@@@id | %user_id%                        |
+      | quota@@@state     | normal                           |
+      | root@@@id         | %space_id%                       |
+      | root@@@webDavUrl  | %base_url%/dav/spaces/%space_id% |
+      | webUrl            | %base_url%/f/%space_id%          |
+
+
+  Scenario: normal user tries to get hid/her own drive information
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:75]:
```suggestion
  Scenario: normal user gets his/her own drive information
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  I want to be able to retrieve user information
+  So that I can see the information
+
+  Background:
+    Given user "Alice" has been created with default attributes and without skeleton files
+    And the administrator has given "Alice" the role "Admin" using the settings api
+
+
+  Scenario: admin user gets the information of a user
+    Given user "Brian" has been created with default attributes and without skeleton files
```
Comment (grgprarup) [inline tests/acceptance/features/apiGraph/getUser.feature:13]: The user `Brian` is created for every scenario, so how about adding this to `Background`?

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    Then the HTTP status code should be "401"
+    And the last response should be an unauthorized response
+
+
+  Scenario: admin user gets all users
+    Given these users have been created with default attributes and without skeleton files:
+      | username |
+      | Brian    |
+      | Carol    |
+      | David    |
```
Comment (grgprarup) [inline tests/acceptance/features/apiGraph/getUser.feature:33]: IMO we don't need to create these two users, `Carol` and `David`. what do you say? We can check for `Admin` and `Alice` users.

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    When user "Alice" tries to get all user using the Graph API
+    Then the HTTP status code should be "200"
+    And the API response should contain all user with following information:
+      | displayName  | id        | mail              | onPremisesSamAccountName |
+      | Brian Murphy | %uuid_v4% | brian@example.org | Brian                    |
+      | David Lopez  | %uuid_v4% | david@example.org | David                    |
+      | Carol King   | %uuid_v4% | carol@example.org | Carol                    |
```
Comment (grgprarup) [inline tests/acceptance/features/apiGraph/getUser.feature:40]:
```suggestion
      | Admin        | %uuid_v4% | admin@example.org | admin                    |
      | Alice Hansen | %uuid_v4% | alice@example.org | Alice                    |
      | Brian Murphy | %uuid_v4% | brian@example.org | Brian                    |
```
IMO instead of creating new users, we can add checks for existing users.

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 14]
```diff
@@ -1240,4 +1245,200 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
 			}
 		}
 	}
+
+	/**
+	 * @When user :user tries to get information of user :ofUser using Graph API
+	 *
+	 * @param string $user
+	 * @param string $ofUser
```
Comment (grgprarup) [inline tests/acceptance/features/bootstrap/GraphContext.php:1253]: We have some already implemented steps, using variables like `byUser` for an acting user and `user` for a target user. It will be better if we use consistent variable names.

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 15]
```diff
@@ -1240,4 +1245,200 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+			}
+			if (!$found) {
+				throw new Exception('User ' . $expectedValue["displayName"] . ' could not be found in the response.');
+			}
+		}
+	}
+
+	/**
+	 * @param string $user
+	 * @param string|null $ofUser
```
Comment (grgprarup) [inline tests/acceptance/features/bootstrap/GraphContext.php:1316]:
```suggestion
	 * @param string $byUser
	 * @param string|null $user
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 16]
```diff
@@ -1240,4 +1245,200 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+	 *
+	 * @return ResponseInterface
+	 * @throws JsonException
+	 * @throws GuzzleException
+	 */
+	public function retrieveUserInformationAlongWithDriveUsingGraphApi(
+		string $user,
+		?string $ofUser = null
```
Comment (grgprarup) [inline tests/acceptance/features/bootstrap/GraphContext.php:1324]:
```suggestion
		string $byUser,
		?string $user = null
```

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 15 above)
Comment (grgprarup) [reply tests/acceptance/features/bootstrap/GraphContext.php:1316]: There are other places too where similar changes are needed.

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 17]
```diff
@@ -1240,4 +1245,200 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+	):ResponseInterface {
+		if ($ofUser === null) {
+			$ofUser = $user;
+		}
```
Comment (grgprarup) [inline tests/acceptance/features/bootstrap/GraphContext.php:1328]:
```suggestion
		$user = $user ?? $byUser;
```
I prefer `null coalescing operator`

File: tests/TestHelpers/GraphHelper.php [hunk 18]
```diff
@@ -242,6 +260,32 @@ public static function deleteUser(
 		);
 	}
 
+	/**
+	 * @param string $baseUrl
+	 * @param string $xRequestId
+	 * @param string $user
```
Comment (grgprarup) [inline tests/TestHelpers/GraphHelper.php:266]:
```suggestion
	 * @param string $byUser
```

File: tests/TestHelpers/GraphHelper.php [hunk 19]
```diff
@@ -242,6 +260,32 @@ public static function deleteUser(
...
+	 *
+	 * @return ResponseInterface
+	 * @throws GuzzleException
+	 */
+	public static function getUserWithDriveInformation(
+		string $baseUrl,
+		string $xRequestId,
+		string $user,
+		string $userPassword,
+		?string $ofUser = null
```
Comment (grgprarup) [inline tests/TestHelpers/GraphHelper.php:278]:
```suggestion
		?string $user = null
```

File: tests/TestHelpers/GraphHelper.php [hunk 20]
```diff
@@ -242,6 +260,32 @@ public static function deleteUser(
...
+	): ResponseInterface {
+		$url = self::getFullUrl($baseUrl, 'users/' . $ofUser . '?%24select=&%24expand=drive');
```
Comment (grgprarup) [inline tests/TestHelpers/GraphHelper.php:280]:
```suggestion
		$url = self::getFullUrl($baseUrl, 'users/' . $user . '?%24select=&%24expand=drive');
```

File: tests/TestHelpers/GraphHelper.php [hunk 21]
```diff
@@ -242,6 +260,32 @@ public static function deleteUser(
...
+		return HttpRequestHelper::get(
+			$url,
+			$xRequestId,
+			$user,
```
Comment (grgprarup) [inline tests/TestHelpers/GraphHelper.php:284]:
```suggestion
			$byUser,
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+  Background:
+    Given these users have been created with default attributes and without skeleton files:
+      | username |
+      | Alice    |
+      | Brian    |
+    And the administrator has given "Alice" the role "Admin" using the settings api
+
+
+  Scenario: admin user gets the information of a user
+    When user "Alice" tries to get information of user "Brian" using Graph API
```
Comment (grgprarup) [inline tests/acceptance/features/apiGraph/getUser.feature:16]:
```suggestion
    When user "Alice" gets information of user "Brian" using Graph API
```
As per the scenario, this will be much more logical. Note: You don't need to implement it separately. We can add it to the same implementation.

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+
+
+  Scenario: non-admin user tries to get the information of a user
+    When user "Brian" tries to get information of user "Alice" using Graph API
+    Then the HTTP status code should be "401"
+    And the last response should be an unauthorized response
+
+
+  Scenario: admin user gets all users
+    When user "Alice" tries to get all user using the Graph API
```
Comment (grgprarup) [inline tests/acceptance/features/apiGraph/getUser.feature:30]:
```suggestion
    When user "Alice" gets all user using the Graph API
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+      | name              | Brian Murphy                     |
+      | owner@@@user@@@id | %user_id%                        |
+      | quota@@@state     | normal                           |
+      | root@@@id         | %space_id%                       |
+      | root@@@webDavUrl  | %base_url%/dav/spaces/%space_id% |
+      | webUrl            | %base_url%/f/%space_id%          |
+
+
+  Scenario: normal user gets his/her own drive information
+    When the user "Brian" tries to get his drive information using Graph API
```
Comment (grgprarup) [inline tests/acceptance/features/apiGraph/getUser.feature:63]:
```suggestion
    When the user "Brian" gets his drive information using Graph API
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 25]
```diff
@@ -1240,4 +1245,198 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
 			}
 		}
 	}
+
+	/**
+	 * @When user :byUser tries to get information of user :user using Graph API
```
Comment (grgprarup) [inline tests/acceptance/features/bootstrap/GraphContext.php:1250]:
```suggestion
	 * @When user :byUser tries to get information of user :user using Graph API
	  * @When user :byUser gets information of user :user using Graph API
```
We can do multiple steps with a single implementation. Only, if you use the suggested version in the feature file.

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 26]
```diff
@@ -1240,4 +1245,198 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+			$this->featureContext->getStepLineRef(),
+			$credentials["username"],
+			$credentials["password"],
+			$user
+		);
+	}
+
+	/**
+	 *
+	 * @When /^the user "([^"]*)" tries to get user "([^"]*)" along with (his|her) drive information using Graph API$/
```
Comment (grgprarup) [inline tests/acceptance/features/bootstrap/GraphContext.php:1339]:
```suggestion
	 * @When /^the user "([^"]*)" tries to get user "([^"]*)" along with (his|her) drive information using Graph API$/
	 * @When /^the user "([^"]*)" gets user "([^"]*)" along with (his|her) drive information using Graph API$/
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 27]
```diff
@@ -1240,4 +1245,198 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+						1,
+						GraphHelper::isUUIDv4($actualKeyValueFromDriveInformation),
+						__METHOD__ .
+						' Expected user_id to have UUIDv4 pattern but found: ' . $actualKeyValueFromDriveInformation
+					);
+					break;
+				case '%space_id%':
+					Assert::assertEquals(
+						1,
+						GraphHelper::isSpaceId($actualKeyValueFromDriveInformation),
```
Comment (grgprarup) [inline tests/acceptance/features/bootstrap/GraphContext.php:1405]:
```suggestion
						GraphHelper::isSpaceId($tempDriveInfo),
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 28]
```diff
@@ -1240,4 +1245,198 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+						__METHOD__ .
+						' Expected user_id to have UUIDv4 pattern but found: ' . $actualKeyValueFromDriveInformation
```
Comment (grgprarup) [inline tests/acceptance/features/bootstrap/GraphContext.php:1407]:
```suggestion
						' Expected user_id to have UUIDv4 pattern but found: ' . $tempDriveInfo
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 29]
```diff
@@ -1240,4 +1245,198 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+						[
+							[
+								"code" => "%space_id%",
+								"function" =>
+									[$this, "getSpaceIdFromActualDriveinformation"],
+								"parameter" => [$actualValueDriveInformation]
+							],
+						]
+					);
+					Assert::assertEquals($expectedDriveInformation[$keyName], $actualKeyValueFromDriveInformation);
```
Comment (grgprarup) [inline tests/acceptance/features/bootstrap/GraphContext.php:1424]:
```suggestion
					Assert::assertEquals($expectedDriveInformation[$keyName], $tempDriveInfo);
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 30]
```diff
@@ -1240,4 +1245,197 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+			$credentials['username'],
+			$credentials['password'],
+			$user
+		);
+		$this->featureContext->setResponse($response);
+	}
+
+	/**
+     * @When user :user tries to get all user using the Graph API
+	 * @When user :user gets all user using the Graph API
```
Comment (kiranparajuli589) [inline tests/acceptance/features/bootstrap/GraphContext.php:1272]:
```suggestion
         * @When user :user tries to get all users using the Graph API
	 * @When user :user gets all users using the Graph API
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 31]
```diff
@@ -1240,4 +1245,197 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+			$this->featureContext->getBaseUrl(),
+			$this->featureContext->getStepLineRef(),
+			$credentials['username'],
+			$credentials['password'],
+		);
+		$this->featureContext->setResponse($response);
+	}
+
+	/**
+	 * @Then the API response should contain all user with following information:
```
Comment (kiranparajuli589) [inline tests/acceptance/features/bootstrap/GraphContext.php:1290]:
```suggestion
	 * @Then the API response should contain all users with the following information:
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 32]
```diff
@@ -1240,4 +1245,197 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+	 * @return void
+	 */
+	public function checkUserDriveInformation(array $expectedDriveInformation, array  $actualValueDriveInformation):void {
+		foreach (array_keys($expectedDriveInformation) as $keyName) {
+			// break the segment with @@@  to find the actual value from the drive infromation from response
+			$separatedKey = explode("@@@", $keyName);
+			// this stores the actual value of each key from drive information response used for assertion
+			$actualKeyValueFromDriveInformation = $actualValueDriveInformation;
+
+			foreach ($separatedKey as $key) {
```
Comment (kiranparajuli589) [inline tests/acceptance/features/bootstrap/GraphContext.php:1389]: the for loop is just assigning different values to `$actualKeyValueFromDriveInformation`. why is that necessary?

Comment (SagarGi) [reply tests/acceptance/features/bootstrap/GraphContext.php:1389]: we have `owner@@@user@@@id` such key passed we need to get actual `id` from it so it is separated and looped.

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 33]
```diff
@@ -1240,4 +1245,197 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+				case '%space_id%':
+					Assert::assertEquals(
+						1,
+						GraphHelper::isSpaceId($actualKeyValueFromDriveInformation),
+						__METHOD__ .
+						' Expected user_id to have UUIDv4 pattern but found: ' . $actualKeyValueFromDriveInformation
+					);
+					break;
+				default:
+					$expectedDriveInformation[$keyName] = $this->featureContext->substituteInLineCodes(
```
[the remark this answers is not in the document]
Comment (SagarGi) [reply tests/acceptance/features/bootstrap/GraphContext.php:1410]: wanted to substitute for value `%base_url%/dav/spaces/%space_id%`. so created separate function.

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 34]
```diff
@@ -1240,4 +1245,200 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+	public function checkUserDriveInformation(array $expectedDriveInformation, array  $actualValueDriveInformation):void {
+		foreach (array_keys($expectedDriveInformation) as $keyName) {
+			// break the segment with @@@  to find the actual value from the drive infromation from response
+			$separatedKey = explode("@@@", $keyName);
+			// this stores the actual value of each key from drive information response used for assertion
+			$actualKeyValueFromDriveInformation = $actualValueDriveInformation;
+
+			foreach ($separatedKey as $key) {
+				$actualKeyValueFromDriveInformation = $actualKeyValueFromDriveInformation[$key];
+			}
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:1393]: Or can be like this
```suggestion
			$actualValue = null;

			foreach ($separatedKey as $key) {
				$actualValue = $actualValueDriveInformation[$key];
			}
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 35]
```diff
@@ -1240,4 +1245,200 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+
+	/**
+	 * check if single drive information is correct
+	 *
+	 * @param array $expectedDriveInformation
+	 * @param array $actualValueDriveInformation
+	 *
+	 * @return void
+	 */
+	public function checkUserDriveInformation(array $expectedDriveInformation, array  $actualValueDriveInformation):void {
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:1384]: what do you think?
```suggestion
	public function checkUserDriveInformation(array $expectedProperties, array  $actualProperties):void {
```

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 34 above)
Comment (kiranparajuli589) [reply tests/acceptance/features/bootstrap/GraphContext.php:1393]: doesn't this for loop sets the last key value in the `$actualValue`. i do not understand the usage of a for loop just to do that.

Comment (kiranparajuli589) [reply tests/acceptance/features/bootstrap/GraphContext.php:1393]: or else the whole switch case assertion should be inside the for loop

File: tests/TestHelpers/GraphHelper.php [hunk 36]
```diff
@@ -26,16 +26,34 @@ private static function getRequestHeaders(): array {
...
 	}
 
+	/**
+	 * @param string $spaceId
+	 *
+	 * @return int (1 = true | 0 = false)
+	 */
+	public static function isSpaceId(string $spaceId): int {
+		$regex = "/^" . self::getUUIDv4Regex() . '\\$' . self::getUUIDv4Regex() . "$/i";
+		return preg_match($regex, $spaceId);
```
Comment (kiranparajuli589) [inline tests/TestHelpers/GraphHelper.php:54]:
```suggestion
		return preg_match($regex, $spaceId) === 1;
```
return bools rather integers, are easier to deal with.

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 33 above)
[the remark this answers is not in the document]
Comment (kiranparajuli589) [reply tests/acceptance/features/bootstrap/GraphContext.php:1413]: add a nice comment so that would be easier to understand for others.

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 37]
```diff
@@ -1240,4 +1245,200 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+		foreach (array_keys($expectedDriveInformation) as $keyName) {
+			// break the segment with @@@  to find the actual value from the drive infromation from response
+			$separatedKey = explode("@@@", $keyName);
```
Comment (kiranparajuli589) [inline tests/acceptance/features/bootstrap/GraphContext.php:1387]: imo maybe doing this in a separate helper func. is more useful.

File: tests/TestHelpers/GraphHelper.php (hunk 36 above)
Comment (saw-jan) [reply tests/TestHelpers/GraphHelper.php:54]: Or cast the type as preg_match will return either `0`, `1` or `false` ([preg_match](https://www.php.net/manual/en/function.preg-match.php))
```php
return (bool)preg_match($regex, $spaceId)
```

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 35 above)
Comment (SagarGi) [reply tests/acceptance/features/bootstrap/GraphContext.php:1384]: i think it about drive so `expectedDriveInformation` would be good

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 37 above)
Comment (SagarGi) [reply tests/acceptance/features/bootstrap/GraphContext.php:1382]: `checkUserDriveInformation` it self a function. separating it would not be reused for others as it is used by this function it self. in doubt what shall i do.

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 35 above)
Comment (saw-jan) [reply tests/acceptance/features/bootstrap/GraphContext.php:1379]: Yeah :+1: , but change `actualValueDriveInformation` to `actualDriveInformation`

**Review by grgprarup — COMMENTED** [review]
lgtm :+1:

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 38]
```diff
@@ -1240,4 +1240,191 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+
+	/**
+	 * check if single drive information is correct
+	 *
+	 * @param array $expectedDriveInformation
+	 * @param array $actualValueDriveInformation
+	 *
+	 * @return void
+	 */
+	public function checkUserDriveInformation(array $expectedDriveInformation, array  $actualValueDriveInformation):void {
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:1379]:
```suggestion
	 * @param array $actualDriveInformation
	 *
	 * @return void
	 */
	public function checkUserDriveInformation(array $expectedDriveInformation, array  $actualDriveInformation):void {
```

File: tests/TestHelpers/GraphHelper.php [hunk 39]
```diff
@@ -26,14 +26,32 @@ private static function getRequestHeaders(): array {
 		];
 	}
 
+	/**
+	 *
+	 * @return string
+	 */
+	public static function getUUIDv4Regex(): string {
+		return '[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}';
```
Comment (kiranparajuli589) [inline tests/TestHelpers/GraphHelper.php:34]: why returning string,,,the func. name says regex.
```suggestion
		return '/^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/';
```
this way you can avoid string concatenation like in L43.

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 40]
```diff
@@ -1240,4 +1240,191 @@ public function checkUserInformation(array $expectedValue, array  $actualValue):
...
+	 * @param TableNode $table
+	 *
+	 * @Then the user retrieve API response should contain the following drive information:
+	 *
+	 * @return void
+	 */
+	public function theResponseShouldContainTheFollowingDriveInformation(TableNode $table): void {
+		$expectedDriveInformation = $table->getRowsHash();
+		// array of user drive information (Personal Drive Information Only)
+		$actualDriveInformation = $this->featureContext->getJsonDecodedResponse($this->featureContext->getResponse())['drive'];
```
Comment (kiranparajuli589) [inline tests/acceptance/features/bootstrap/GraphContext.php:1427]: check first the `$this->featureContext->getJsonDecodedResponse($this->featureContext->getResponse())` statement is array and has `drive` key in it.. will be more easier for debugging in case of malformed responses.

File: tests/TestHelpers/GraphHelper.php (hunk 39 above)
Comment (saw-jan) [reply tests/TestHelpers/GraphHelper.php:34]: Needed this to build other regex patterns like in L53 for space id

**Review by saw-jan — COMMENTED** [review]
LGTM

**Review by micbar — APPROVED** [review]
please add tests always first on `stable-2.0` and forward port to master.

**Comment by SagarGi** [comment]
> please add tests always first on `stable-2.0` and forward port to master.

sorry @micbar this PR was actually created way before back porting thing. Will be doing for up coming PRs.  :smile:

File: tests/TestHelpers/GraphHelper.php [hunk 41]
```diff
@@ -26,14 +26,54 @@ private static function getRequestHeaders(): array {
...
+	 * Key name can consist of @@@
+	 * This function separate such key and return its actual value from actual drive response which can be used for assertion
+	 *
+	 * @param string $keyName
+	 * @param array $actualDriveInformation
+	 *
+	 * @return string
+	 */
+	public static function separateAndGetValueForKey(string $keyName, array $actualDriveInformation): string {
+		// break the segment with @@@  to find the actual value from the actual drive infromation
```
Comment (phil-davis) [inline tests/TestHelpers/GraphHelper.php:67]:
```suggestion
		// break the segment with @@@  to find the actual value from the actual drive information
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 42]
```diff
@@ -0,0 +1,77 @@
...
+  Scenario: admin user gets all users
+    When user "Alice" gets all users using the Graph API
+    Then the HTTP status code should be "200"
+    And the API response should contain all users with the following information:
+      | displayName  | id        | mail              | onPremisesSamAccountName |
+      | Alice Hansen | %uuid_v4% | alice@example.org | Alice                    |
+      | Brian Murphy | %uuid_v4% | brian@example.org | Brian                    |
+
+
+  Scenario: non-admin user tries get all user
```
Comment (phil-davis) [inline tests/acceptance/features/apiGraph/getUser.feature:38]:
```suggestion
  Scenario: non-admin user tries to get all users
```

**Review by phil-davis — APPROVED** [review]

**Comment by phil-davis** [comment]
I rebased to get up-to-date CI. This can be merged when CI passes.

**Comment by phil-davis** [comment]
Also ported to `stable-2.0` in PR #5311
